lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/AirCompressorCodec.java

```diff
@@ -0,0 +1,140 @@
+    public static final Supplier<AirCompressorCodec> SNAPPY_CODEC_SUPPLIER = () -> new AirCompressorCodec(DataFileConstants.SNAPPY_CODEC, new SnappyCompressor(), new SnappyDecompressor(), bytes -> SnappyDecompressor.getUncompressedLength(bytes, 0), true);
+
+    public static final CodecFactory SNAPPY_CODEC_FACTORY = new CodecFactory()
+    {
+        @Override
+        protected Codec createInstance()
+        {
+            return SNAPPY_CODEC_SUPPLIER.get();
+        }
+    };
+
+    @VisibleForTesting
+    public static final Supplier<AirCompressorCodec> ZSTD_CODEC_SUPPLIER = () -> new AirCompressorCodec(DataFileConstants.ZSTANDARD_CODEC, new ZstdCompressor(), new ZstdDecompressor(), bytes -> {
+        int decompressedSize = toIntExact(ZstdDecompressor.getDecompressedSize(bytes, 0, bytes.length));
+        if (decompressedSize < 0) {
+            return bytes.length * 2;
+        }
+        return decompressedSize;
+    }, false);
+
+    public static final CodecFactory ZSTD_CODEC_FACTORY = new CodecFactory()
+    {
+        @Override
+        protected Codec createInstance()
+        {
+            return ZSTD_CODEC_SUPPLIER.get();
+        }
+    };
+
+    private final String name;
+    private final Compressor compressor;
+    private final Decompressor decompressor;
+    private final Function<byte[], Integer> decompressionSizeGeneration;
+    private final boolean includeCrc32;
+    private final CRC32 crc32 = new CRC32();
+
+    public AirCompressorCodec(String name, Compressor compressor, Decompressor decompressor, Function<byte[], Integer> decompressionSizeGeneration, boolean includeCrc32)
+    {
+        this.name = requireNonNull(name, "name is null");
+        this.compressor = requireNonNull(compressor, "compressor is null");
+        this.decompressor = requireNonNull(decompressor, "decompressor is null");
+        this.decompressionSizeGeneration = requireNonNull(decompressionSizeGeneration, "decompressionSizeGeneration is null");
+        this.includeCrc32 = includeCrc32;
+    }
+
+    @Override
+    public String getName()
+    {
+        return name;
+    }
+
+    @Override
+    public ByteBuffer compress(ByteBuffer uncompressedData)
+            throws IOException
+    {
+        if (includeCrc32) {
+            crc32.reset();
+            crc32.update(uncompressedData.array(), uncompressedData.arrayOffset() + uncompressedData.position(), uncompressedData.remaining());
+        }
+        ByteBuffer output = ByteBuffer.allocate(compressor.maxCompressedLength(uncompressedData.remaining()) + (includeCrc32 ? 4 : 0));
+        compressor.compress(uncompressedData, output);
+        if (includeCrc32) {
+            output.putInt((int) crc32.getValue());
+        }
+        output.flip();
+        return output;
+    }
+
+    @Override
+    public ByteBuffer decompress(ByteBuffer compressedData)
+            throws IOException
+    {
+        int remaining = compressedData.remaining() - (includeCrc32 ? 4 : 0);
```

Comment:
Should we validate the CRC here? The Avro Snappy codec validates it.

Reply:
just added before I saw this. Do it below.